src/mono/mono/mini/jit-icalls.c

```diff
@@ -94,7 +95,10 @@ ldvirtfn_internal (MonoObject *obj, MonoMethod *method, gboolean gshared)
 
 	/* An rgctx wrapper is added by the trampolines no need to do it here */
 
-	return mono_ldftn (res);
+	addr =  mono_ldftn (res);
+	if (m_class_is_valuetype (mono_object_class(obj)))
```

Comment:
I don't think this check is correct. We shouldn't add unbox wrapper for instance methods other than the ones coming from the valuetype's special parents (like System.ValueType). Only there we are passing a boxed object. For normal value type method calls, `this` is a managed pointer, not a boxed value type.

Reply:
Another problem is that the unbox trampoline will wrap 'addr' which is itself a trampoline, so every time its called, the call will go through the whole trampoline in common_call_trampoline (). So it will probably work, but it will be very slow.

Reply:
The correct should be 
`if (res->klass is valuetype && !method->klass !s valuetype) => add unbox tramp to res`